wnfs-common/src/blockstore.rs

```diff
@@ -8,134 +6,181 @@ use libipld::{
+    let second_loaded: Vec<u8> = store.get_deserializable(&second_cid).await.unwrap();
 
-    /// Retrieves an array of bytes from the block store with given CID.
-    async fn get_block<'a>(&'a self, cid: &Cid) -> Result<Cow<'a, Vec<u8>>> {
-        let bytes = self
-            .0
-            .get(&cid.to_string())
-            .ok_or(BlockStoreError::CIDNotFound(*cid))?;
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(first_loaded, first_bytes);
+    assert_eq!(second_loaded, second_bytes);
 
-        Ok(Cow::Borrowed(bytes))
-    }
+    // Return Ok
+    Ok(())
 }
 
-//--------------------------------------------------------------------------------------------------
-// Functions
-//--------------------------------------------------------------------------------------------------
+// Generic function used to test any type that conforms to the BlockStore trait
+pub async fn bs_duplication<T: BlockStore + Send + 'static>(store: &T) -> Result<()> {
+    // Example objects to insert and remove from the blockstore
+    let first_bytes = vec![1, 2, 3, 4, 5];
+    let second_bytes = first_bytes.clone();
+
+    // Insert the objects into the blockstore
+    let first_cid = store.put_serializable(&first_bytes).await.unwrap();
+    let second_cid = store.put_serializable(&second_bytes).await.unwrap();
+
+    // Assert that the two vecs produced the same CID
+    assert_eq!(first_cid, second_cid);
+
+    // Retrieve the objects from the blockstore
+    let first_loaded: Vec<u8> = store.get_deserializable(&first_cid).await.unwrap();
+    let second_loaded: Vec<u8> = store.get_deserializable(&second_cid).await.unwrap();
+
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(first_loaded, first_bytes);
+    assert_eq!(second_loaded, second_bytes);
+    // Assert that the objects we loaded are the same
+    assert_eq!(first_loaded, second_loaded);
+
+    // Return Ok
+    Ok(())
+}
+
+pub async fn bs_serialization<
+    T: BlockStore + Send + Serialize + 'static + for<'de> Deserialize<'de>,
+>(
+    store: &T,
+) -> Result<()> {
+    // Example objects to insert and remove from the blockstore
+    let bytes = vec![1, 2, 3, 4, 5];
+    // Insert the object into the blockstore
+    let cid = store.put_serializable(&bytes).await.unwrap();
+
+    // Create a buffer to hold the serialized BlockStore
+    let mut writer: Vec<u8> = Vec::new();
+    // Serialize the blockstore
+    serde_json::to_writer_pretty(&mut writer, &store).unwrap();
+
+    // Construct a new BlockStore from the Serialized object
+    let new_store: T = serde_json::from_reader(writer.as_slice()).unwrap();
+
+    // Retrieve the object from the blockstore
+    let loaded: Vec<u8> = new_store.get_deserializable(&cid).await.unwrap();
+
+    // Assert that the objects are the same as the ones we inserted
+    assert_eq!(loaded, bytes);
+    Ok(())
+}
 
 #[cfg(test)]
-mod tests {
+pub mod tests {
```

Comment:
Not a comment about this PR, but more generally:
@appcypher I think this pattern will continue to crop up where a library exposes utilities for testing code based on top of it, similar to how we have the test-strategies feature for the wnfs (and wnfs-hamt?) crate. Should we give that flag a more generic name, like "test-utils" and feature-gate this export with that flag, too? Are you aware of other libraries doing this as a pattern?

Reply:
Oh actually the `pub` on the `mod tests` here can be removed, you likely moved the `bs_duplication` etc. functions out, right Vera?

Reply:
yes, you're right!
will fix